src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java

```diff
@@ -13,61 +13,82 @@
  */
 package org.apache.hadoop.fs;
 
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableSet;
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.UserGroupInformation.AuthenticationMethod;
 import org.apache.hadoop.util.ReflectionUtils;
 
 import java.io.IOException;
-import java.lang.reflect.Field;
 import java.net.URI;
+import java.util.HashMap;
 import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
 import java.util.concurrent.atomic.AtomicLong;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.util.Objects.requireNonNull;
 import static org.apache.hadoop.fs.FileSystem.getFileSystemClass;
+import static org.apache.hadoop.security.UserGroupInformationShim.getSubject;
 
 public final class PrestoFileSystemCache
         extends FileSystem.Cache
 {
     private static final int CACHE_LEAK_PROTECTION_THRESHOLD = Integer.parseInt(System.getProperty("hadoopFileSystemCacheLeakProtectionThreshold", "100000"));
 
     private final AtomicLong unique = new AtomicLong();
-    private final Map<Key, FileSystem> map;
-
-    @SuppressWarnings("unchecked")
-    public PrestoFileSystemCache()
-    {
-        map = getPrivateMap(this);
-    }
+    private final Map<FileSystemKey, FileSystemHolder> map = new HashMap<>();
 
     @Override
     FileSystem get(URI uri, Configuration conf)
             throws IOException
     {
-        return getInternal(uri, conf, new Key(uri, conf));
+        return getInternal(uri, conf, 0);
     }
 
     @Override
     FileSystem getUnique(URI uri, Configuration conf)
             throws IOException
     {
-        return getInternal(uri, conf, new Key(uri, conf, unique.incrementAndGet()));
+        return getInternal(uri, conf, unique.getAndIncrement());
```

Comment:
This needs to be `incrementAndGet()`. Otherwise, if you call `get()` followed by `getUnique()`, they'll both have `0` and return the same FS.